src/repositories/shopping-cart.repository.ts

```diff
@@ -4,14 +4,69 @@
 // License text available at https://opensource.org/licenses/MIT
 
 import {DefaultKeyValueRepository} from '@loopback/repository';
-import {ShoppingCart} from '../models/shopping-cart.model';
+import {ShoppingCart, ShoppingCartItem} from '../models/shopping-cart.model';
 import {RedisDataSource} from '../datasources/redis.datasource';
 import {inject} from '@loopback/context';
+import {promisify} from 'util';
 
 export class ShoppingCartRepository extends DefaultKeyValueRepository<
   ShoppingCart
 > {
   constructor(@inject('datasources.redis') ds: RedisDataSource) {
     super(ShoppingCart, ds);
   }
+
+  /**
+   * Add an item to the shopping cart with optimistic lock to allow concurrent
+   * `adding to cart` from multiple devices
+   *
+   * @param userId User id
+   * @param item Item to be added
+   */
+  addItem(userId: string, item: ShoppingCartItem) {
+    const addItemToCart = (cart: ShoppingCart | null) => {
+      cart = cart || new ShoppingCart({userId});
+      cart.items = cart.items || [];
+      cart.items.push(item);
+      return cart;
+    };
+    return this.checkAndSet(userId, addItemToCart);
+  }
+
+  /**
+   * Use Redis WATCH and Transaction to check and set against a key
+   * See https://redis.io/topics/transactions#optimistic-locking-using-check-and-set
```

Comment:
@raymondfeng IIUC, the algorithm outlined in the linked page is expecting to re-run the transaction if a race condition was detected, and repeat those re-run until the transaction succeed.

I don't see that logic implemented here. Are we expecting the REST clients to retry the operation? Could you please explain how are we signaling the race condition to them?